tests/big_bird/test_modeling_big_bird.py

```diff
@@ -596,6 +596,15 @@ def test_for_change_to_full_attn(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_for_change_to_full_attn(*config_and_inputs)
 
+    # overwrite from common in order to skip the check on `attentions`
+    def check_outputs(self, fx_outputs, pt_outputs, model_class, names):
+        # `bigbird_block_sparse_attention` in `FlaxBigBird` returns `attention_probs = None`, while in PyTorch version,
+        # an effort was done to return `attention_probs` (yet to be verified).
+        if type(names) == str and names.startswith("attentions"):
+            return
+        else:
+            super().check_outputs(fx_outputs, pt_outputs, model_class, names)
```

Comment:
Not really in favor of using `super()` here. This makes the test very difficult to understand. I'd prefer to just copy paste the code here. 

cc @sgugger @LysandreJik

Reply:
Strong disagree for me. The test will likely diverge from the original at the next change/fix in it. Also, I don't see what's hard to understand about calling the superclass method.

The test code is different from the modeling code, and we should make it as easy as possible for us to maintain.

Reply:
I am also in much favor of not using copy for the testing (unless in the cases where a call to parent class is not easy)
- the philosophy and design choice are mainly for models and for users (that would/need to read the architectures)
- we make our effort to maintain the library in this way, but it is not very easy (hope I am not the only one :sweat_smile:)
- the code divergences between models could be detected by (well implemented) tests - if those divergences are indeed bugs
- however, the divergences of testing codes are difficult to notice and to sync
- and testing codes are mainly for our internal usage + for (very) advanced users to add new models --> they probably get used to such design patterns (maybe they will prefer it too!)

Let's wait for opinions from @LysandreJik and @patil-suraj too!

Reply:
I agree with @ydshieh here. For testing IMO it should be fine to use inheritance since it's for internal use. It's also easy to maintain tests this way.

Reply:
Don't agree here - would like to discuss this a bit :-) Thanks for the arguments so far! Here is my opinion.

We previously said that testing code is VIP and should have the same standard as non-testing code. Why do we make a difference here in terms of readability and code philosophy?

If we support such design choices from now on we will strongly diverge from our fundamental self-contained file philosophy here. What's different here? The testing code should be as inviting for community contributions as the non-testing code.
Here are my arguments:

- **Readability**: The test is now much harder to read. Imagine a community contributor now starts changing BigBird and this test fails. She/he then has to follow this `super()` statement to jump into the `modeling_common_tests.py` file, will adapt the test there. Then 100+ test will fail.
- **Invites for hacky fixes** Everybody will start adding hacks to `test_modeling_common.py` to make it work for their specific use case. We're already doing this now with `self.has_attentions` which I don't like. Some tests in `test_modeling_common.py` have already become unreadable for users IMO, *e.g.* the `test_attentions.py` test in `test_modeling_common.py`
- **Philosohpy mismatch** Think it's also confusing in terms of philosophy to allow abstractions in tests but not in modeling code

The main points I disagree with from @ydshieh and @patil-suraj are

```
and testing codes are mainly for our internal usage + for (very) advanced users to add new models --> they probably get used to such design patterns (maybe they will prefer it too!
``` 
=> This is very much not True IMO. Why are tests just for internal use? Every contributor has to look into the tests. There is not a single PR where one would run a test to verify it or where we wouldn't ask to add a test. The first thing I do when I don't understand a model or function in another github code base is that I start looking into the test.

The test diverge can be easily solved in my opinion by using the `# Copied from ...` methodology IMO. It would also fit well with the model design. E.g. Wav2Vec2 was the first model, so Wav2Vec2's tests were the first ones as well. Then HuBERT's tests can be "# Copied from" from Wav2Vec2's tests.

But more generally whichever principle we decide to follow here we need to agree on the same design IMO. Otherwise, I continue overwriting test functions and other continue adding `abstractions` and new functionality to general tests, which is not a good idea I think :sweat_smile: 

For me it would be ideal if we only have general tests in `test_modeling_common.py` that (without hacks) apply to **all** models. I would then also be fine with adding very large `TestMixins` for model classes, such as generative models or maybe vision and speech models, but I'm not really in favor of using `super()` inside sub functions that are then used in another test. If one debugs the check equivalence tests there is lots of jumping around between files I think.

Reply:
Nice arguments! I personally would favor the user of `super()` in this situation. Testing code is indeed VIP, and we should definitely have a very high standard in terms of both readability and correctness, however, I don't think we should start to apply the same code philosophy that we're currently applying for models to the testing code:

- This philosophy emerges from the want to have independent modeling files for each model, so that they're easy to modify, easy to tweak without affecting other models, and easily customizable/monkey-patchable as their internals are exposed. I think this cannot, and should not, be applied to the testing code -> the testing code should be as centralized as possible in the common testing files, in order to ensure that all models have a similar behavior and a similar API.
- I also disagree that tests are not for internal use; tests should be contributed by the community, and the aim is really not to limit that. It's instead to offer a simple way to override tests while ensuring that it will not diverge. I personally think expecting users to read the entire content of the common testing files and their accompanying mixins puts a very very high bar for contributions, whereas it should be very simple to add the mixin, run the tests, see what fails and update accordingly.
- I also don't think it's bad that we have two different philosophies regarding two very different parts of the library. I think it's perfectly fine to have different expectations to the testing code, which ultimately will always fall on the maintainers to ensure that it is correctly maintained and implemented.

Finally, I agree with you Patrick that it would be ideal to have very general tests in `test_modeling_common.py`, and then largish mixins for the different modalities/specificities that you mention (generative, vision, speech, attention-less, etc). This is, in my opinion, the best way to have true model agnostic APIs while ensuring that all modalities are covered sufficiently, in a reproducible manner. But I don't think this means we shouldn't override tests with `super` when there are a few changes in the test to make it pass (which shouldn't happen in most cases, but it's unavoidable in others).